File: data_structures/linked_lists/ll_queue/python/linkedListQueue.py
class Node:
    def __init__(self, val=None, nextNode=None):
        self.val = val
        self.next = nextNode

class Queue:
    def __init__(self, length = 0, head = None, tail = None):
        self.length = length
        self.head = head
        self.tail = tail 
    
    def enqueue(self, node):
        """enqueue"""
        self.length += 1
        if self.tail == None:
            self.tail = node
            self.head = node
        else:
            self.tail.next = node
            self.tail = node

    def dequeue(self):
        """dequeue"""
        if self.head == None:
            return None
        else:
            self.length -= 1
            head = self.head
            self.head = self.head.next
            head.next = None
            return head.val

    def peek(self):
        """peek at the first element in the queue"""
        return self.head
    
if __name__ == "__main__":
    q = Queue()
    print(q.peek())
    q.enqueue(Node(4, None))
    print(q.peek().val)
    q.enqueue(Node(5, None))
    print(q.peek().val)
    q.dequeue()
    print(q.peek().val)

    

File: data_structures/linked_lists/ll_queue/java/linkedListQueue.java
import java.util.LinkedList;

class LinkedListQueue {
    Node head;
    Node tail;
    int length = 0;

    public void queue(int val)
    {
        this.length++;
        Node new_node = new Node(val);
        if(head == null)
        {
            this.head = new_node;
            this.tail = new_node;
        }
        else
        {
            this.tail.next = new_node;
            this.tail = new_node;
        }
    }

    public void dequeue()
    {
        if(this.length == 0)
        {
            System.out.println("The queue is empty");
        }
        else if (this.length == 1)
        {
            System.out.println(this.head.val);
            this.head = null;
            this.tail = null;
            this.length--;
        }
        else
        {
            System.out.println(this.head.val);
            this.head = this.head.next;
        }
    }

    public void peek()
    {
        if( this.head == null)
        {
            System.out.println("Queue is empty");
        }
        else
        {
            System.out.println(this.head.val);
        }
    }

    public static void main(String[] args)
    {
        LinkedListQueue linked_list = new LinkedListQueue();
        linked_list.queue(5);
        linked_list.peek(); //five has entered the queue and should be at front
        linked_list.queue(4);
        linked_list.queue(3);
        linked_list.queue(2);
        linked_list.queue(1);
        linked_list.peek(); //five should still be at the front
        linked_list.dequeue();
        linked_list.dequeue();
        linked_list.peek(); //three should be at the front of the queue
        linked_list.dequeue();
        linked_list.dequeue();
        linked_list.peek(); //one should be at the front of the queue
        linked_list.dequeue();
        linked_list.peek(); //queue should be at the front of the queue

    }
}
